crates/uv/src/commands/project/add.rs

```diff
@@ -33,10 +40,93 @@ pub(crate) async fn add(
     // Find the project requirements.
     let project = ProjectWorkspace::discover(&std::env::current_dir()?, None).await?;
 
+    // Discover or create the virtual environment.
+    let venv = project::init_environment(project.workspace(), python.as_deref(), cache, printer)?;
+
+    let client_builder = BaseClientBuilder::new()
+        .connectivity(connectivity)
+        .native_tls(native_tls)
+        .keyring(settings.keyring_provider);
+
+    // Read the requirements.
+    let RequirementsSpecification { requirements, .. } =
+        RequirementsSpecification::from_sources(&requirements, &[], &[], &client_builder).await?;
+
+    // TODO(charlie): These are all default values. We should consider whether we want to make them
+    // optional on the downstream APIs.
+    let exclude_newer = None;
+    let python_version = None;
+    let python_platform = None;
+    let hasher = HashStrategy::default();
+    let setup_py = SetupPyStrategy::default();
+    let config_settings = ConfigSettings::default();
+    let build_isolation = BuildIsolation::default();
+
+    // Use the default settings.
+    let settings = ResolverSettings::default();
+
+    // Determine the environment for the resolution.
+    let (tags, markers) =
+        resolution_environment(python_version, python_platform, venv.interpreter())?;
+
+    // Initialize the registry client.
+    let client = RegistryClientBuilder::new(cache.clone())
```

Comment:
Created #4330 to track the duplication of this after declaring a `BaseClientBuilder` above.